### PR TITLE
De-Regression: eagerly evaluate not_implemented_for in decorated generators

### PR DESCRIPTION
**Pull request by boothby**
See issue #5520 for details.

~Work in progress, needs doc updates.~

~Also, there's an issue where `open_file` and `not_implemented_for` are like oil and water -- specifically `not_implemented_for` is like oil and should go on top.  This isn't currently an issue in the codebase but should be addressed before it causes an obnoxious bug (see #5520 for discussion)~

### CODE REVIEW

File: networkx/utils/tests/test_decorators.py [hunk 1]
```diff
@@ -489,3 +489,29 @@ def nop(self):
...
+    def test_immediate_raise(self):
+        @not_implemented_for("directed")
+        def yield_nodes(G):
+            yield from G
+
+        G = nx.Graph([(1, 2)])
+        D = nx.DiGraph()
+
+        print([*yield_nodes(G)])
+        print(yield_nodes._code)
```
Comment (rossbar) [inline networkx/utils/tests/test_decorators.py:502]: I have a slight preference for swapping out the print function with something more test-y, e.g. in the first case maybe something like:

```
assert list(yield_nodes(G)) == [1, 2]
```

I don't fully grok what the second print tests, but if it's just the existence of the `_code` attr, maybe replace with an `assert hasattr`?

**Review by rossbar — APPROVED** [review]
LGTM, with a nit about prints in the tests. The added documentation+example for eager mapping is excellent, thanks @boothby !

File: networkx/utils/tests/test_decorators.py (hunk 1 above)
Comment (boothby) [reply networkx/utils/tests/test_decorators.py]: Good catch; the prints were only meant to be temporary.

**Comment by dschult** [comment]
It seems that now the decorator is not eager for generators using `yield from G` but still is eager for generators using ```for v in G: yield v```.  Something subtle going on....??

**Comment by boothby** [comment]
What I see happening is that it's lazy for the first call only, and I see the boneheaded logic that's causing the behavior.  I paused for a moment's indecision; @rossbar's suggestion of simply forbidding try-finally decorations of generators is quite appealing.  I'll whip that up today.

File: networkx/utils/tests/test_decorators.py [hunk 2]
```diff
@@ -485,7 +481,27 @@ def test_indent(self):
...
+    def test_immediate_raise(self):
+        @not_implemented_for("directed")
+        def yield_nodes(G):
+            yield from G
+
+        G = nx.Graph([(1, 2)])
+        D = nx.DiGraph()
+
+        list(yield_nodes(G))
+        print(yield_nodes._code)
```
Comment (dschult) [inline networkx/utils/tests/test_decorators.py:493]: Maybe remove print statement and change first call to `yield_nodes` into a test?  
But the next comment suggests that it is testing the "first call". so maybe the `list(yield_nodes(G))` line should be removed?  Not sure...
```suggestion
        assert list(yield_nodes(G)) == [1, 2]
```

Comment (boothby) [reply networkx/utils/tests/test_decorators.py]: *sigh* missed that delta... gone again

**Review by dschult — APPROVED** [review]
This looks good!   Thanks @boothby !!

**Pull request merged**